### DISCUSSION TITLE
User IDs are invertible through exhaustive search

### DISCUSSION THREAD

**Discussion by ott** [plausible/analytics#1152] [Anything Else]
## Prerequisites
- [x] I have searched open and closed issues to make sure that the bug has not yet been reported.

## Bug report
**Describe the bug**
The algorithm to generate a user ID (unique visitor ID) is not a practical one-way function. It can be easily inverted through exhaustive search.

This is in particular the case for IPv4 addresses. Theoretically, the IPv4 address space has a cardinality of 2<sup>32</sup>. In practice it is even lower due to hierarchical address space sub-divisions, unused and reserved address space and other factors. Moreover, there have also been successful enumeration attempts of larger parts of the IPv6 address space, for example, through routing table analysis, predictably assigned IPv6 addresses or predictable MAC addresses. Moreover, both the IPv6 and IPv4 address space can be enumerated through techniques such as passive traffic observation, routing table analysis and active probing through advertisements.

The number of common user agent strings is also finite and low. Both domain and hostname are known. Therefore, an attacker or even just a user who is in possession of the salt can easily enumerate the all user IDs of IPv4 addresses through exhaustive search. The search space should have a cardinality of roughly 2<sup>40</sup> in theory for full coverage and less than 2<sup>20</sup> in practice for good coverage. Therefore, such an enumeration attack is practical with a single personal computer and without any noteworthy hardware or energy costs (from the perspective of an industrialized country).

I was not able to find information on the pre-image resistance of SipHash but perhaps it is possible use properties of SipHash speed up the exhaustive search any further as SipHash seems to be designed as a MAC and not a cryptographic hash function. However, given the low cardinality of the search space it does not even seem to be necessary to use any property of SipHash.

**Expected behavior**
User IDs are not invertible through exhaustive search.

**Additional remarks**
It might be impossible though to find an algorithm does not use session IDs, cookies, similar storage mechanism or IP addresses.

I could not find a concrete threat model for the software, so perhaps this enumeration and de-anonymization attack is outside of the assumed threat model and does therefore not constitute an attack because it is assumed to be impossible.

**Comment by ukutaht** [comment]
Thanks for the discussion. I agree that IP Adress and User-Agent by themselves do not have enough entropy to make the hash irreversible. This is why the salt is used and later deleted.

Are we making the claim somewhere that the hash is irreversible even when you're in possession of the salt? That would be incorrect. Our stance that the hash is irreversible relies on the random salt being used and then deleted.
